resolwe/flow/models/annotations.py

```diff
@@ -0,0 +1,243 @@
+        for vocabulary_value, vocabulary_label in self.vocabulary.items():
+            if vocabulary_label == label:
+                return vocabulary_value
+        return label
+
+    @property
+    def annotation_type(self) -> AnnotationType:
+        """Get the field type as enum."""
+        return AnnotationType(self.type)
+
+    @staticmethod
+    def group_field_from_path(path: str) -> List[str]:
+        """Return the group and field name from path."""
+        return path.split(".", maxsplit=1)
+
+    @classmethod
+    def id_from_path(cls, path: str) -> Optional[int]:
+        """Get the field id from the field path."""
+        group_name, field_name = cls.group_field_from_path(path)
+        return (
+            cls.objects.filter(group__name=group_name, name=field_name)
+            .values_list("id", flat=True)
+            .first()
+        )
+
+    @classmethod
+    def field_from_path(cls, path: str) -> Optional["AnnotationField"]:
+        """Get the field id from the field path."""
+        group_name, field_name = cls.group_field_from_path(path)
+        return cls.objects.filter(group__name=group_name, name=field_name).first()
+
+    def __str__(self) -> str:
+        """Return user-friendly string representation."""
+        return f"{self.group.name}.{self.name}"
+
+    class Meta:
+        """Set the constraints and the default ordering."""
+
+        constraints = [
+            # Accept only supported annotation types.
+            models.constraints.CheckConstraint(
+                check=models.Q(type__in=[e.value for e in AnnotationType]),
+                name="annotation_type",
+            ),
+            models.constraints.UniqueConstraint(
+                fields=["name", "group"], name="uniquetogether_name_group"
+            ),
+        ]
+        ordering = ["group__sort_order", "sort_order"]
+
+
+class AnnotationPreset(AuditModel, PermissionObject):
+    """The named set of annotation fields.
+
+    The preset have permissions.
+    """
+
+    #: the name of the annotation fields
+    name = models.CharField(max_length=128)
+
+    #: the fields belongint to this preset
+    fields = models.ManyToManyField(AnnotationField, related_name="presets")
+
+    #: the sorting order among annotation fields
+    sort_order = models.PositiveSmallIntegerField(default=1)
+
+    class Meta:
+        """Override definition in base classes."""
+
+
+class AnnotationValue(AuditModel):
+    """The value of the annotation."""
+
+    #: the entity this field belongs to
+    entity: "Entity" = models.ForeignKey(
+        "Entity", related_name="annotations", on_delete=models.CASCADE
+    )
+
+    #: the field this field belongs to
+    field: AnnotationField = models.ForeignKey(
+        AnnotationField, related_name="values", on_delete=models.CASCADE
+    )
+
+    #: value is stored under key 'value' in the json field to simplify lookups
+    json: Any = models.JSONField(default={})
+
+    @property
+    def value(self):
```

Comment:
This is actually a total mess... I would prefer a single JSON value, but the dictionary makes it nicer to use the value in dynamic lookups in the vocabulary dictionary. I will try to find a way to get rid of this completely, otherwise, I will follow your suggestion.

Reply:
I have implemented your suggestion.